### PR TITLE
feat(table-permission): support table level authority control (beta)

### LINKED ISSUE #1585: [Feature]: More fine-grained table access control
### Is your feature request related to a problem?

1. I want to allow different members of the project to have specific permissions on a specific tables. But this is not possible under the current authority control system.

### Describe the solution you'd like

Allow different members in the project to have specific permissions on specific tables;
More fine-grained permission classification, such as: data query permission, data update permission, DDL permission, data export permission, etc.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ISADBA**
#### What type of PR is this?
add database owner role
add table permission control

#### Which issue(s) this PR fixes:

Fixes #1585 #1679

#### Special notes for your reviewer:
have frontend change

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by smallsheeeep — COMMENTED** [review]
Thanks for your controbutions! 😆
I have reviewed part of the code and suggested changes. The rest of the code will be reviewed later.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowTaskInstanceService.java [hunk 1]
```diff
@@ -255,6 +256,8 @@ public List<? extends FlowTaskResult> getTaskResultFromEntity(@NotNull TaskEntit
             return getApplyProjectResult(taskEntity);
         } else if (taskEntity.getTaskType() == TaskType.APPLY_DATABASE_PERMISSION) {
             return getApplyDatabaseResult(taskEntity);
+        } else if (taskEntity.getTaskType() == TaskType.APPLY_TABLE_PERMISSION) {
+            return getApplyTableResult(taskEntity);
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowTaskInstanceService.java:260]: Apply table permission task has no file for download, so please add `instance.getTaskType() != TaskType.APPLY_TABLE_PERMISSION` in `FlowTaskInstanceService#getDownloadableTaskEntity()`.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/iam/model/UserResourceRole.java [hunk 2]
```diff
@@ -44,6 +44,8 @@ public class UserResourceRole implements PermissionConfiguration {
 
     private ResourceRoleName resourceRole;
 
+    private Long resourceRoleId;
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/iam/model/UserResourceRole.java:47]: Delete this useless codes.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/iam/ResourceRoleService.java [hunk 3]
```diff
@@ -222,4 +222,17 @@ private UserResourceRole fromEntity(UserResourceRoleEntity entity, ResourceRoleE
         return model;
     }
 
+    public List<UserResourceRole> getUserIdsByResourceIdAndTypeAndName(Long resourceId, ResourceType resourceType,
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/iam/ResourceRoleService.java:225]: This method could also be deleted.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/iam/UserService.java [hunk 4]
```diff
@@ -395,6 +395,12 @@ public User detail(long id) {
         return detailWithoutPermissionCheck(id);
     }
 
+    @SkipAuthorize("odc internal usage")
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/iam/UserService.java:398]: This method could also be deleted.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/permission/database/DatabasePermissionHelper.java [hunk 5]
```diff
@@ -60,6 +63,12 @@ public class DatabasePermissionHelper {
     @Autowired
     private UserDatabasePermissionRepository userDatabasePermissionRepository;
 
+    @Autowired
+    private UserTablePermissionRepository userTablePermissionRepository;
+
+    @Autowired
+    private TablePermissionService tablePermissionService;
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/permission/database/DatabasePermissionHelper.java:70]: Delete this useless codes.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/model/ApplyTableParameter.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    public static class ApplyTable implements Serializable {
+        private static final long serialVersionUID = -8433967513537417701L;
+
+        private Long databaseId;
+        private String databaseName;
+        private Long dataSourceId;
+        private String dataSourceName;
+        private Long tableId;
+        private String tableName;
+        private List<String> tableNames;
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/model/ApplyTableParameter.java:74]: Why we need both `tableName` and `tableNames`? Which fields in `ApplyTable` are filled by the front-end?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/model/QueryTablePermissionParams.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+public class QueryTablePermissionParams {
+    private Long userId;
+    private Long ticketId;
+    private String fuzzyDatabaseName;
+    private Long databaseId;
+    private String fuzzyDataSourceName;
+    private Long dataSourceId;
+    private List<DatabasePermissionType> types;
+    private AuthorizationType authorizationType;
+    private List<ExpirationStatusFilter> statuses;
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/model/QueryTablePermissionParams.java:46]: Query by `fuzzyTableName` is needed.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java [hunk 8]
```diff
@@ -0,0 +1,604 @@
...
+    @Autowired
+    private DatabasePermissionHelper databasePermissionHelper;
+
+    @Value("${odc.iam.permission.expired-retention-time-seconds:7776000}")
+    private long expiredRetentionTimeSeconds;
+
+    private static final UserTablePermissionMapper mapper = UserTablePermissionMapper.INSTANCE;
+    private static final int EXPIRING_DAYS = 7;
+
+    private static final Set<String> ORACLE_DATA_DICTIONARY = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java:143]: `ORACLE_DATA_DICTIONARY` and `MYSQL_DATA_DICTIONARY` are always empty?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java [hunk 9]
```diff
@@ -0,0 +1,604 @@
...
+        }
+        Date expiredTime = new Date(System.currentTimeMillis() - expiredRetentionTimeSeconds * 1000);
+        permissionService.deleteExpiredPermission(expiredTime);
+        Date expireTimeThreshold = TimeUtils.getStartOfDay(new Date());
+        Specification<UserTablePermissionEntity> spec = Specification
+                .where(UserTablePermissionSpec.projectIdEqual(projectId))
+                .and(UserTablePermissionSpec.organizationIdEqual(authenticationFacade.currentOrganizationId()))
+                .and(UserTablePermissionSpec.userIdEqual(params.getUserId()))
+                .and(UserTablePermissionSpec.ticketIdEqual(params.getTicketId()))
+                .and(UserTablePermissionSpec.databaseNameLike(params.getFuzzyDatabaseName()))
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java:165]: Please add `tableNameLike` Spec.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java [hunk 10]
```diff
@@ -0,0 +1,604 @@
...
+                    } else {
+                        permission.setStatus(ExpirationStatusFilter.NOT_EXPIRED);
+                    }
+                    return permission;
+                });
+    }
+
+    @Transactional(rollbackFor = Exception.class)
+    @SkipAuthorize("permission check inside")
+    public List<UserTablePermission> listWithoutPage(@NotNull Long projectId,
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java:187]: This method could be deleted. If you want to query table permission and the result without page, you can invoke `list(projectId, params, Pageable.unpaged()).getContent();`

File: server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java [hunk 11]
```diff
@@ -0,0 +1,604 @@
...
+                    } else {
+                        permission.setStatus(ExpirationStatusFilter.NOT_EXPIRED);
+                    }
+                    return permission;
+                }).collect(Collectors.toList());
+    }
+
+    @Transactional(rollbackFor = Exception.class)
+    @SkipAuthorize("odc internal usage")
+    public List<UserTablePermission> listWithoutPageByDataSourceId(@NotNull Long dataSourceId,
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java:225]: This method could be deleted too. you can use the `list()` above, and for some no need parameters, just set `null`.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java [hunk 12]
```diff
@@ -0,0 +1,604 @@
...
+                    } else {
+                        permission.setStatus(ExpirationStatusFilter.NOT_EXPIRED);
+                    }
+                    return permission;
+                }).collect(Collectors.toList());
+    }
+
+    @Transactional(rollbackFor = Exception.class)
+    @PreAuthenticate(hasAnyResourceRole = {"OWNER", "DBA"}, resourceType = "ODC_PROJECT", indexOfIdParam = 0)
+    public List<UserTablePermission> batchCreate(@NotNull Long projectId,
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java:252]: Can we merge these two method `batchCreate` , `taskBatchCreate`?

**Review by smallsheeeep — COMMENTED** [review]
This is really a huge project, and I haven’t reviewed a PR with so much code in a long time. The following are new PR modification suggestions.

File: libs/db-browser/src/main/resources/schema/sql/mysql/mysql_5_7_x.yaml [hunk 13]
```diff
@@ -140,7 +140,7 @@ sqls:
       any_value(SUBPARTITION_EXPRESSION) as SUBPARTITION_EXPRESSION,
       any_value(PARTITION_DESCRIPTION) as PARTITION_DESCRIPTION
     FROM
-      information_schema.partitions
+      information_schema.`partitions`
```
[the remark this answers is not in the document]
Comment (ISADBA) [reply libs/db-browser/src/main/resources/schema/sql/mysql/mysql_5_7_x.yaml:143]: in doris datasource, partitions is a keyword, will be take syntax error.

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_16__add_resource_role.yaml [hunk 14]
```diff
@@ -0,0 +1,17 @@
+kind: resource
```
[the remark this answers is not in the document]
Comment (ISADBA) [reply server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_16__add_resource_role.yaml:1]: done

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_21__alter_iam_permission.sql [hunk 15]
```diff
@@ -0,0 +1,5 @@
+alter table `iam_permission` add column `resource_type` varchar(64) default null comment 'ResourceType: ODC_DATABASE,ODC_TABLE,ODC_COLUMN ';
```
[the remark this answers is not in the document]
Comment (ISADBA) [reply server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_21__alter_iam_permission.sql:1]: done

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_21__alter_iam_permission.sql [hunk 16]
```diff
@@ -0,0 +1,5 @@
...
+alter table `iam_permission` add column `resource_id` bigint(20) default null comment 'resource id';
+alter table `iam_permission` add key `idx_resource_type_and_id` (resource_type,resource_id);
+update `iam_permission` set `resource_type` = 'ODC_DATABASE' where `resource_identifier` LIKE 'ODC_DATABASE:%'  and `action` in ('query', 'change', 'export');
+update `iam_permission` set `resource_id` = SUBSTRING( `resource_identifier` FROM 14 ) where `resource_type` = 'ODC_DATABASE';
```
[the remark this answers is not in the document]
Comment (ISADBA) [reply server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_21__alter_iam_permission.sql:5]: done

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_22__add_list_user_table_permission_view.sql [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+      i_p.`expire_time` as `expire_time`,
+      i_p.`creator_id` as `creator_id`,
+      i_p.`organization_id` as `organization_id`
+    from
+      `iam_permission` as i_p
+      inner join `iam_user_permission` as i_up on i_p.`id` = i_up.`permission_id`
+    where
+      i_p.`resource_type` = 'ODC_DATABASE'
+  ) as u_p
+  inner join `connect_database` as c_d on SUBSTRING( u_p.`resource_identifier` FROM 14 ) = c_d.`id`
```
[the remark this answers is not in the document]
Comment (ISADBA) [reply server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_22__add_list_user_table_permission_view.sql:40]: done

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_22__add_list_user_table_permission_view.sql [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+  inner join `connect_connection` as c_c on c_d.`connection_id` = c_c.`id`;
+
+--
+-- Add `list_user_table_permission_view` view for query user database permissions list
+--
+
+
+
+create or replace view `list_user_table_permission_view` as
```
[the remark this answers is not in the document]
Comment (ISADBA) [reply server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_22__add_list_user_table_permission_view.sql:49]: Please fix this issue when merging into the v4.3 branch.

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_20__add_connect_table.sql [hunk 19]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS `connect_table` (
```
[the remark this answers is not in the document]
Comment (ISADBA) [reply server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_20__add_connect_table.sql:1]: Please fix this issue when merging into the v4.3 branch.

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_22__add_list_user_table_permission_view.sql [hunk 20]
```diff
@@ -0,0 +1,72 @@
+--
```
[the remark this answers is not in the document]
Comment (ISADBA) [reply server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_22__add_list_user_table_permission_view.sql:1]: done

File: server/odc-core/src/main/resources/i18n/BusinessMessages_zh_TW.properties [hunk 21]
```diff
@@ -202,7 +203,8 @@ com.oceanbase.odc.AuditEventAction.CREATE_PARTITION_PLAN_TASK=創建分區計劃
 com.oceanbase.odc.AuditEventAction.CREATE_ALTER_SCHEDULE_TASK=創建修改調度任務
 com.oceanbase.odc.AuditEventAction.CREATE_ONLINE_SCHEMA_CHANGE_TASK=創建無鎖結構變更任務
 com.oceanbase.odc.AuditEventAction.CREATE_APPLY_PROJECT_PERMISSION_TASK=創建申請項目權限任務
-com.oceanbase.odc.AuditEventAction.CREATE_APPLY_DATABASE_PERMISSION_TASK=創建申請數據庫權限任務
+com.oceanbase.odc.AuditEventAction.CREATE_APPLY_DATABASE_PERMISSION_TASK=創建申請數據表權限任務
```
[the remark this answers is not in the document]
Comment (ISADBA) [reply server/odc-core/src/main/resources/i18n/BusinessMessages_zh_TW.properties:206]: done

File: server/odc-core/src/main/resources/i18n/BusinessMessages.properties [hunk 22]
```diff
@@ -833,6 +843,7 @@ com.oceanbase.odc.notification.channel-test-message=【ODC】Channel verificatio
 #
 com.oceanbase.odc.builtin-resource.permission-apply.project.description=Requesting [{1}] permissions for project [{0}]
 com.oceanbase.odc.builtin-resource.permission-apply.database.description=Requesting [{0}] permissions for databases
+com.oceanbase.odc.builtin-resource.permission-apply.table.description=Requesting [{0}] permissions for database’s tables
```
[the remark this answers is not in the document]
Comment (ISADBA) [reply server/odc-core/src/main/resources/i18n/BusinessMessages.properties:846]: done

File: server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/TablePermissionController.java [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+    @ApiOperation(value = "listTablePermissions", notes = "List table permissions")
+    @RequestMapping(value = "", method = RequestMethod.GET)
+    public PaginatedResponse<UserTablePermission> list(@PathVariable Long projectId,
+            @RequestParam(name = "userId", required = false) Long userId,
+            @RequestParam(name = "ticketId", required = false) Long ticketId,
+            @RequestParam(name = "databaseName", required = false) String fuzzyDatabaseName,
+            @RequestParam(name = "dataSourceName", required = false) String fuzzyDataSourceName,
+            @RequestParam(name = "type", required = false) List<DatabasePermissionType> types,
+            @RequestParam(name = "authorizationType", required = false) AuthorizationType authorizationType,
+            @RequestParam(name = "status", required = false) List<ExpirationStatusFilter> statuses,
```
[the remark this answers is not in the document]
Comment (ISADBA) [reply server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/TablePermissionController.java:68]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/connection/ConnectionEntity.java [hunk 24]
```diff
@@ -79,7 +79,7 @@ public class ConnectionEntity {
     private ConnectionVisibleScope visibleScope;
 
     /**
-     * Owner id, user_id if visible_scope=PRIVATE, enterprise_id if visible_scope=ORGANIZATION
+     * DatabaseOwner id, user_id if visible_scope=PRIVATE, enterprise_id if visible_scope=ORGANIZATION
```
[the remark this answers is not in the document]
Comment (ISADBA) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/connection/ConnectionEntity.java:82]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java (hunk 9 above)
Comment (ISADBA) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java:165]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java (hunk 8 above)
Comment (ISADBA) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java:143]: fixed

Comment (ISADBA) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java:143]: add default value for dictionary, and modify ignoreDataDictionary  default is false.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java (hunk 10 above)
Comment (ISADBA) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java:187]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java (hunk 11 above)
Comment (ISADBA) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java:225]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java (hunk 12 above)
Comment (ISADBA) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/permission/table/TablePermissionService.java:252]: done

**Pull request merged**